bin/inx-chronicle/src/stardust_inx/syncer.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::VecDeque, ops::Range};
+
+use async_trait::async_trait;
+use chronicle::{
+    db::MongoDb,
+    runtime::{Actor, ActorContext, ActorError, ConfigureActor, HandleEvent, Report},
+};
+use inx::{client::InxClient, tonic::Channel};
+
+use super::{InxError, MilestoneStream};
+
+// The Syncer goes backwards in time and tries collect as many milestones as possible.
+pub struct Syncer {
+    gaps: Gaps,
+    db: MongoDb,
+    inx_client: InxClient<Channel>,
+}
+
+impl Syncer {
+    pub fn new(gaps: Vec<Range<u32>>, db: MongoDb, inx_client: InxClient<Channel>) -> Self {
+        Self {
+            gaps: Gaps(gaps.into()),
+            db,
+            inx_client,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Gaps(VecDeque<Range<u32>>);
+
+impl Iterator for Gaps {
+    type Item = Range<u32>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(range) = self.0.pop_front() {
+                if range.start >= range.end {
+                    continue;
+                } else {
+                    break Some(range);
+                }
+            } else {
+                break None;
+            }
+        }
```

Comment:
```suggestion
        while let Some(range) = self.0.pop_front() {
            if range.start < range.end {
                return Some(range);
            }
        }
        None
```

Reply:
Good call 😆. Not sure why I made this so complicated.